nifi-nar-bundles/nifi-framework-bundle/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/engine/StandardStatelessEngine.java

```diff
@@ -0,0 +1,550 @@
+
+        extensionManager.logClassLoaderDetails();
+
+        // Create a child group and add it to the root group. We do this, rather than interacting with the root group directly
+        // because the flow may well have Local Input/Output ports, and those are not allowed on the Root Group.
+        final ProcessGroup rootGroup = flowManager.getRootGroup();
+        final ProcessGroup childGroup = flowManager.createProcessGroup("stateless-flow");
+        childGroup.setName("Stateless Flow");
+        rootGroup.addProcessGroup(childGroup);
+
+        childGroup.updateFlow(dataflowDefinition.getFlowSnapshot(), "stateless-component-id-seed", false, true, true);
+
+        // Map existing parameter contexts by name
+        final Set<ParameterContext> parameterContexts = flowManager.getParameterContextManager().getParameterContexts();
+        final Map<String, ParameterContext> parameterContextMap = parameterContexts.stream()
+            .collect(Collectors.toMap(ParameterContext::getName, context -> context));
+
+        // Update Parameters to match those that are provided in the flow configuration, plus those overrides provided
+        final List<ParameterContextDefinition> parameterContextDefinitions = dataflowDefinition.getParameterContexts();
+        if (parameterContextDefinitions != null) {
+            parameterContextDefinitions.forEach(contextDefinition -> registerParameterContext(contextDefinition, parameterContextMap));
+        }
+
+        overrideParameters(parameterContextMap, parameterOverrides);
+
+        final List<ReportingTaskNode> reportingTaskNodes = createReportingTasks(dataflowDefinition);
+        final StandardStatelessFlow dataflow = new StandardStatelessFlow(childGroup, reportingTaskNodes, controllerServiceProvider, processContextFactory,
+            repositoryContextFactory, dataflowDefinition);
+        dataflow.initialize(processScheduler);
+        return dataflow;
+    }
+
+    private void loadNecessaryExtensions(final DataflowDefinition<VersionedFlowSnapshot> dataflowDefinition) {
+        final VersionedProcessGroup group = dataflowDefinition.getFlowSnapshot().getFlowContents();
+        final Set<BundleCoordinate> requiredBundles = gatherRequiredBundles(group);
+
+        for (final ReportingTaskDefinition reportingTaskDefinition : dataflowDefinition.getReportingTaskDefinitions()) {
+            final BundleCoordinate coordinate = parseBundleCoordinate(reportingTaskDefinition);
+            if (coordinate == null) {
+                continue;
+            }
+
+            requiredBundles.add(coordinate);
+        }
+
+        final int concurrentDownloads = 4;
+        final ExecutorService executor = new FlowEngine(concurrentDownloads, "Download Extensions", true);
+        final Future<Set<Bundle>> future = extensionRepository.fetch(requiredBundles, executor, concurrentDownloads);
+        executor.shutdown();
+
+        logger.info("Waiting for bundles to complete download...");
```

Comment:
The number of bundles isn't really known at this point. The "initial number" is known - as in, requiredBundles.size(). But then each of those may have dependencies, which results in additional bundles being downloaded. So I avoided logging the number here, as it may end up being more confusing.